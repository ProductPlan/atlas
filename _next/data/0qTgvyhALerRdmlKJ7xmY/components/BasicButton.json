{"pageProps":{"component":"BasicButton"},"__N_SSG":true}