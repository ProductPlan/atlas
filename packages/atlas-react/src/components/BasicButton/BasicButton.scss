@import "../../common/styles/atlas";

.BasicButton.BasicButton--default {
  --#{$namespace}-button-background-color: #{color(blue-400)};
  --#{$namespace}-button-background-color-hover: #{color(blue-500)};
  --#{$namespace}-button-background-color-focus: #{color(blue-700)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(blue-700)};
  --#{$namespace}-button-color: #{color(white-400)};
}

.BasicButton.BasicButton--ghost {
  --#{$namespace}-button-background-color: transparent;
  --#{$namespace}-button-background-color-hover: #{color(light-gray-300)};
  --#{$namespace}-button-background-color-focus: #{color(light-gray-300)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(light-gray-300)};
  --#{$namespace}-button-color: #{color(black-400)};
}

.BasicButton.BasicButton--secondary {
  --#{$namespace}-button-background-color: #{color(blue-100)};
  --#{$namespace}-button-background-color-hover: #{color(blue-200)};
  --#{$namespace}-button-background-color-focus: #{color(blue-200)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(blue-200)};
  --#{$namespace}-button-color: #{color(blue-400)};
}

.BasicButton.BasicButton--tertiary {
  --#{$namespace}-button-background-color: #{color(light-gray-100)};
  --#{$namespace}-button-background-color-hover: #{color(light-gray-400)};
  --#{$namespace}-button-background-color-focus: #{color(light-gray-400)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(black-400)};
  --#{$namespace}-button-color: #{color(black-400)};
}

.BasicButton.BasicButton--destructive {
  --#{$namespace}-button-background-color: #{color(red-400)};
  --#{$namespace}-button-background-color-hover: #{color(red-500)};
  --#{$namespace}-button-background-color-focus: #{color(red-500)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(red-100)};
  --#{$namespace}-button-background-color-active: #{color(red-500)};
  --#{$namespace}-button-color: #{color(white-400)};
}

.BasicButton.BasicButton--size-default  {
  --#{$namespace}-button-padding: #{spacer(mini)} #{spacer(small)};
  --#{$namespace}-button-height: 36px;
  --#{$namespace}-button-font-size: var(--a-size-font-body-14);
  --#{$namespace}-button-width: auto;
  --#{$namespace}-button-leading-icon-margin: 0 #{spacer(mini)} 0 0;
  --#{$namespace}-button-trailing-icon-margin: 0 0 0 #{spacer(mini)};
}

.BasicButton.BasicButton--size-full-width  {
  --#{$namespace}-button-padding: #{spacer(mini)} #{spacer(small)};
  --#{$namespace}-button-height: 36px;
  --#{$namespace}-button-font-size: var(--a-size-font-body-14);
  --#{$namespace}-button-width: 100%;
  --#{$namespace}-button-leading-icon-margin: 0 #{spacer(mini)} 0 0;
  --#{$namespace}-button-trailing-icon-margin: 0 0 0 #{spacer(mini)};
}

.BasicButton.BasicButton--size-small  {
  --#{$namespace}-button-padding: #{spacer(micro)} calc(#{spacer(micro)} + #{spacer(mini)});
  --#{$namespace}-button-height: 28px;
  --#{$namespace}-button-font-size: var(--a-size-font-body-14);
  --#{$namespace}-button-width: auto;
  --#{$namespace}-button-leading-icon-margin: 0 #{spacer(micro)} 0 0;
  --#{$namespace}-button-trailing-icon-margin: 0 0 0 #{spacer(micro)};
}

.BasicButton.BasicButton--size-mini  {
  --#{$namespace}-button-padding: #{spacer(micro)} #{spacer(mini)};
  --#{$namespace}-button-height: 26px;
  --#{$namespace}-button-font-size: var(--a-size-font-body-12);
  --#{$namespace}-button-width: auto;
  --#{$namespace}-button-leading-icon-margin: 0 #{spacer(micro)} 0 0;
  --#{$namespace}-button-trailing-icon-margin: 0 0 0 #{spacer(micro)};
}

.BasicButton.BasicButton--align-left {
  --#{$namespace}-button-align: flex-start;
}

.BasicButton.BasicButton--align-right {
  --#{$namespace}-button-align: flex-end;
}

.BasicButton.BasicButton--align-center {
  --#{$namespace}-button-align: center;
}

.BasicButton {
  & + & {
    margin-left: spacer(small);
  }
}

.BasicButton[disabled] {
  pointer-events: none;
}

.BasicButton {
  @include fontFamily('body');
  padding: var(--#{$namespace}-button-padding);
  background-color: var(--#{$namespace}-button-background-color);
  border-radius: 4px;
  border: none;
  color: var(--#{$namespace}-button-color);
  font-size: var(--#{$namespace}-button-font-size);
  line-height: 1;
  display: inline-flex;
  align-items: center;
  justify-content: var(--#{$namespace}-button-align);
  height: var(--#{$namespace}-button-height);
  cursor: pointer;
  width: var(--#{$namespace}-button-width);

  &:focus {
    outline: none;
    box-shadow: 0px 0px 0px 2px var(--#{$namespace}-button-background-color-focus-shadow);
    background-color: var(--#{$namespace}-button-background-color-focus);
  }

  &:active {
    background-color: var(--#{$namespace}-button-background-color-active);
  }

  &:disabled {
    opacity: 0.5;
  }

  &:hover {
    background-color: var(--#{$namespace}-button-background-color-hover);
  }

  .trailingIcon {
    & > * {
      display: inline-flex;
      justify-content: center;
      align-items: center;
      margin: var(--#{$namespace}-button-trailing-icon-margin);
    }
  }

  .leadingIcon {
    & > * {
      display: inline-flex;
      justify-content: center;
      align-items: center;
      margin: var(--#{$namespace}-button-leading-icon-margin);
    }
  }
}


