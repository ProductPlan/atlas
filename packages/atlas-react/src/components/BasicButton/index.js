export { default, buttonVariants, buttonSizes, buttonAlignments } from "./BasicButton";
