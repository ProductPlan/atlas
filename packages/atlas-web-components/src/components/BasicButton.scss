@import "../styles/atlas.scss";

.BasicButton.BasicButton--default {
  --#{$namespace}-button-background-color: #{color(blue-400)};
  --#{$namespace}-button-background-color-hover: #{color(blue-500)};
  --#{$namespace}-button-background-color-focus: #{color(blue-700)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(blue-700)};
  --#{$namespace}-button-color: #{color(white-400)};
}

.BasicButton.BasicButton--ghost {
  --#{$namespace}-button-background-color: transparent;
  --#{$namespace}-button-background-color-hover: #{color(light-gray-300)};
  --#{$namespace}-button-background-color-focus: #{color(light-gray-300)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(light-gray-300)};
  --#{$namespace}-button-color: #{color(black-400)};
}

.BasicButton.BasicButton--secondary {
  --#{$namespace}-button-background-color: #{color(blue-100)};
  --#{$namespace}-button-background-color-hover: #{color(blue-200)};
  --#{$namespace}-button-background-color-focus: #{color(blue-200)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(blue-200)};
  --#{$namespace}-button-color: #{color(blue-400)};
}

.BasicButton.BasicButton--tertiary {
  --#{$namespace}-button-background-color: #{color(light-gray-100)};
  --#{$namespace}-button-background-color-hover: #{color(light-gray-400)};
  --#{$namespace}-button-background-color-focus: #{color(light-gray-400)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(blue-200)};
  --#{$namespace}-button-background-color-active: #{color(black-400)};
  --#{$namespace}-button-color: #{color(black-400)};
}

.BasicButton.BasicButton--destructive {
  --#{$namespace}-button-background-color: #{color(red-400)};
  --#{$namespace}-button-background-color-hover: #{color(red-500)};
  --#{$namespace}-button-background-color-focus: #{color(red-500)};
  --#{$namespace}-button-background-color-focus-shadow: #{color(red-100)};
  --#{$namespace}-button-background-color-active: #{color(red-500)};
  --#{$namespace}-button-color: #{color(white-400)};
}

.BasicButton {
  padding: spacer(mini) spacer(small);
  background-color: var(--#{$namespace}-button-background-color);
  border-radius: 4px;
  border: none;
  color: var(--#{$namespace}-button-color);
  font-size: 14px;
  line-height: 1;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: 36px;
  cursor: pointer;

  &:focus {
    outline: none;
    box-shadow: 0px 0px 0px 2px var(--#{$namespace}-button-background-color-focus-shadow);
    background-color: var(--#{$namespace}-button-background-color-focus);
  }

  &:active {
    background-color: var(--#{$namespace}-button-background-color-active);
  }

  &:disabled {
    opacity: 0.5;
  }

  &:hover {
    background-color: var(--#{$namespace}-button-background-color-hover);
  }
}

.trailingIcon {
  height: 20px;
  width: 20px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  margin-right: -8px;
  margin-left: 4px;
}

.leadingIcon {
  height: 20px;
  width: 20px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  margin-left: -8px;
  margin-right: 4px;
}

