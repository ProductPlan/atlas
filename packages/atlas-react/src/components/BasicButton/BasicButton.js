import React from "react";
import PropTypes from "prop-types";
import "./BasicButton.scss";

const buttonHtmlTypes = ["button", "submit", "reset"];
export const buttonVariants = [
  "default",
  "secondary",
  "tertiary",
  "ghost",
  "destructive",
];
export const buttonSizes = ["default", "small", "mini", "full-width"];
export const buttonAlignments = ["left", "center", "right"];

/** A Basic Button */
export default function BasicButton({
  label,
  htmlId,
  onClick,
  align = "center",
  disabled = undefined,
  htmlType = "button",
  leadingIcon = null,
  size = "default",
  trailingIcon = null,
  type = "default",
}) {
  return (
    <button
      className={`BasicButton BasicButton--${type} BasicButton--size-${size} BasicButton--align-${align}`}
      type={htmlType}
      disabled={disabled}
      onClick={onClick}
      id={htmlId}
    >
      {leadingIcon && <span className="leadingIcon">{leadingIcon}</span>}
      {label}
      {trailingIcon && <span className="trailingIcon">{trailingIcon}</span>}
    </button>
  );
}

BasicButton.propTypes = {
  /** Button content alignment (left, center, right) */
  align: PropTypes.oneOf(buttonAlignments),
  /** The button is disabled or not */
  disabled: PropTypes.bool,
  /** Text to display in the button */
  label: PropTypes.string.isRequired,
  /** html type attribute, one of button, submit, or reset */
  htmlType: PropTypes.oneOf(buttonHtmlTypes),
  /** html id attribute */
  htmlId: PropTypes.string,
  /** one of default, secondary, tertiary, destructive, ghost */
  type: PropTypes.oneOf(buttonVariants),
  /** Click Handler */
  onClick: PropTypes.func,
  /** button size (default, small, mini, full-width) */
  size: PropTypes.oneOf(buttonSizes),
  /** icon to display to the right of the label */
  trailingIcon: PropTypes.element,
  /** icon to display to the left of the label */
  leadingIcon: PropTypes.element,
};
