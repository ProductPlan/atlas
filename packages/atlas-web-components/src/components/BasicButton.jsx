import { define } from 'preactement';
import "redefine-custom-elements";
import "./BasicButton.scss";

const BasicButton = ({
  align = "center",
  disabled = undefined,
  htmlId = undefined,
  label = "",
  leadingIcon = null,
  size = "default",
  theme = "default",
  trailingIcon = null,
  type = "button",
}) => {
  return (
    <>
      <button
        className={`BasicButton BasicButton--${theme} BasicButton--size-${size} BasicButton--align-${align}`}
        type={type}
        htmlId={htmlId}
        disabled={disabled}
        size={size}
      >
        {leadingIcon && <span className="leadingIcon">{leadingIcon}</span>}
        {label}
        {trailingIcon && <span className="trailingIcon">{trailingIcon}</span>}
      </button>
    </>
  );
};

define('atlas-basic-button', () => BasicButton, { attributes: ['align', 'disabled', 'htmlId', 'label', 'size', 'theme', 'type']})
