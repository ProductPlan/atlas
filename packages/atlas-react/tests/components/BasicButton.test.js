import React from "react";
import { render, screen } from "@testing-library/react";
import BasicButton, { buttonVariants } from "@components/BasicButton";

const icon = <i className="fa fa-chevron-left"></i>;

describe("BasicButton", () => {
  it("handles the click onClick handler", () => {
    const onClick = jest.fn();
    render(
      <BasicButton
        label="Click Me"
        onClick={onClick}
        htmlId="buttonWithClickHandler"
      />
    );

    screen.getByText("Click Me").click();
    expect(onClick).toBeCalled();
    expect(onClick).toBeCalledTimes(1);
  });

  it("renders the htmlForm property", () => {
    const htmlTypes = ["button", "submit", "reset"];

    htmlTypes.forEach((type) => {
      const label = `Click Me - ${type}`;
      render(
        <BasicButton
          label={label}
          htmlForm="formId"
          htmlType={type}
          htmlId="buttonWithClickHandler"
        />
      );

      const buttonElement = screen.getByText(label);
      expect(buttonElement.getAttribute("form")).toBe("formId");
      expect(buttonElement.getAttribute("type")).toBe(type);
    });
  });

  it("renders a different htmlType if provided", () => {
    render(
      <BasicButton label="Click Me" htmlType="submit" htmlId="buttonWithForm" />
    );

    expect(screen.getByText("Click Me").getAttribute("type")).toBe("submit");
  });

  describe("when using different variants", () => {
    buttonVariants.forEach((type) => {
      it(`renders a ${type} button`, () => {
        const { container } = render(
          <BasicButton label="Click Me" type={type} htmlId={`button-${type}`} />
        );
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe("when using icons", () => {
    buttonVariants.forEach((type) => {
      describe(`when button with type ${type} has a leading icon`, () => {
        it("renders the icon", () => {
          const { container } = render(
            <BasicButton
              label="Click Me"
              type={type}
              htmlId={`button-${type}`}
              leadingIcon={icon}
            />
          );
          expect(container).toMatchSnapshot();
        });
      });

      describe(`when button with type ${type} has a trailing icon`, () => {
        it("renders the icon", () => {
          const { container } = render(
            <BasicButton
              label="Click Me"
              type={type}
              htmlId={`button-${type}`}
              trailingIcon={icon}
            />
          );
          expect(container).toMatchSnapshot();
        });
      });
    });
  });
});
