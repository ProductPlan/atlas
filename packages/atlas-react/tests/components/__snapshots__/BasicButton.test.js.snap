// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicButton when using different sizes renders a default button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-center"
    id="button-default"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different sizes renders a full-width button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-full-width BasicButton--align-center"
    id="button-full-width"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different sizes renders a mini button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-mini BasicButton--align-center"
    id="button-mini"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different sizes renders a small button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-small BasicButton--align-center"
    id="button-small"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different text alignments renders a center button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-center"
    id="button-center"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different text alignments renders a left button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-left"
    id="button-left"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different text alignments renders a right button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-right"
    id="button-right"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different variants renders a default button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-center"
    id="button-default"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different variants renders a destructive button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--destructive BasicButton--size-default BasicButton--align-center"
    id="button-destructive"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different variants renders a ghost button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--ghost BasicButton--size-default BasicButton--align-center"
    id="button-ghost"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different variants renders a secondary button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--secondary BasicButton--size-default BasicButton--align-center"
    id="button-secondary"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using different variants renders a tertiary button 1`] = `
<div>
  <button
    class="BasicButton BasicButton--tertiary BasicButton--size-default BasicButton--align-center"
    id="button-tertiary"
    type="button"
  >
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type default has a leading icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-center"
    id="button-default"
    type="button"
  >
    <span
      class="leadingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type default has a trailing icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--default BasicButton--size-default BasicButton--align-center"
    id="button-default"
    type="button"
  >
    Click Me
    <span
      class="trailingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type destructive has a leading icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--destructive BasicButton--size-default BasicButton--align-center"
    id="button-destructive"
    type="button"
  >
    <span
      class="leadingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type destructive has a trailing icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--destructive BasicButton--size-default BasicButton--align-center"
    id="button-destructive"
    type="button"
  >
    Click Me
    <span
      class="trailingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type ghost has a leading icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--ghost BasicButton--size-default BasicButton--align-center"
    id="button-ghost"
    type="button"
  >
    <span
      class="leadingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type ghost has a trailing icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--ghost BasicButton--size-default BasicButton--align-center"
    id="button-ghost"
    type="button"
  >
    Click Me
    <span
      class="trailingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type secondary has a leading icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--secondary BasicButton--size-default BasicButton--align-center"
    id="button-secondary"
    type="button"
  >
    <span
      class="leadingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type secondary has a trailing icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--secondary BasicButton--size-default BasicButton--align-center"
    id="button-secondary"
    type="button"
  >
    Click Me
    <span
      class="trailingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type tertiary has a leading icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--tertiary BasicButton--size-default BasicButton--align-center"
    id="button-tertiary"
    type="button"
  >
    <span
      class="leadingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
    Click Me
  </button>
</div>
`;

exports[`BasicButton when using icons when button with type tertiary has a trailing icon renders the icon 1`] = `
<div>
  <button
    class="BasicButton BasicButton--tertiary BasicButton--size-default BasicButton--align-center"
    id="button-tertiary"
    type="button"
  >
    Click Me
    <span
      class="trailingIcon"
    >
      <i
        class="fa fa-chevron-left"
      />
    </span>
  </button>
</div>
`;
